lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,373 @@
+use crate::TxSyncError;
+
+use lightning::util::logger::Logger;
+use lightning::{log_error, log_given_level, log_info, log_internal, log_debug, log_trace};
+use lightning::chain::WatchedOutput;
+use lightning::chain::{Confirm, Filter};
+
+use bitcoin::{BlockHash, BlockHeader, Script, Transaction, Txid};
+
+use esplora_client::Builder;
+#[cfg(feature = "async-interface")]
+use esplora_client::r#async::AsyncClient;
+#[cfg(not(feature = "async-interface"))]
+use esplora_client::blocking::BlockingClient;
+
+use std::sync::Mutex;
+use std::collections::HashSet;
+use core::ops::Deref;
+
+/// Synchronizes LDK with a given [`Esplora`] server.
+///
+/// Needs to be registerd with a [`ChainMonitor`] via the [`Filter`] interface to be informed of
+/// transactions and outputs to montor.
+///
+/// [`Esplora`]: https://github.com/Blockstream/electrs
+/// [`ChainMonitor`]: lightning::chain::chainmonitor::ChainMonitor
+/// [`Filter`]: lightning::chain::Filter
+pub struct EsploraSyncClient<L: Deref>
+where
+	L::Target: Logger,
+{
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	queued_transactions: Mutex<HashSet<Txid>>,
```

Comment:
Yeah, I had also thought about that, but I think just having a bool is not as simple as it seems:

1. We don't want to hold a shared `Mutex` for the entire sync since that would block the world, or at least everything that calls the `Filter` functions.
2. Just having an `AtomicBool` indicating if a sync is pending raises the question when to reset it. If we reset it at the beginning of the round, the sync could still fail. If we reset it at the end, we don't know if there were no new registration during the round.

So I think having queues that are processed on a per-round basis is pretty straighforward here. However, while thinking about this again I realized that we still need a bool *in addition* which indicates if we need to resync right away, e.g., after having encountered an error or inconsistency before. Added this with https://github.com/lightningdevkit/rust-lightning/pull/1870/commits/0a89f9df3295a5f93e3703248f235c5d947c22bd.